### DISCUSSION TITLE
is this possible to convert socket into django channels

### DISCUSSION THREAD

**Discussion by hemanth-sp** [django/channels#1661] [Q&A]
hi,
I am integrating one device that device will send data using TCP protocol to given IP address this https://www.codegrepper.com/code-examples/python/Recieve+and+send+data+with+tcp+sockets+python works fine, using the IP address but how to convert into Django channels?
![Screenshot from 2020-11-27 18-22-38](https://user-images.githubusercontent.com/24698571/100451607-b5be8600-30dd-11eb-81f3-926451a94ced.png)

**Comment by mondwan** [comment]
I am not sure how you want to integrate django channel with your application. I mean I am not sure whether you want to integrate on the sending side (the one sending data using TCP .... to someone else) / receiving side.

* Let's say you want to integrate django channel into the sending side
* Then, you need to also send your message, let's call it MX, (in Channels format) via Channel Layer to your database (redis usually) in your sending side
* Next, you need to code a consumer, let's call it CX, to react on MX accordingly
* Finally, mount your CX to a Worker. You will be good I think.

**Comment by yedpodtrzitko** [comment]
I think there's probably some confusion. TCP connection is [a transport layer (4)](https://en.wikipedia.org/wiki/OSI_model#Layer_4:_Transport_Layer), whereas Websocket/HTTP happens a few layers higher on the [application layer (7)](https://en.wikipedia.org/wiki/OSI_model#Layer_7:_Application_Layer), so you cant "convert" it. You have to build an abstractions layers to make the data structure conforming the Websocket protocol (or whatever Application Layer Protocol you choose).
